drivers/builtin/src/nist_kw.c

```diff
@@ -39,167 +40,140 @@ static const unsigned char NIST_KW_ICV1[] = { 0xA6, 0xA6, 0xA6, 0xA6, 0xA6, 0xA6
 static const  unsigned char NIST_KW_ICV2[] = { 0xA6, 0x59, 0x59, 0xA6 };
 
 /*
- * Initialize context
+ * Helper function for Xoring the uint64_t "t" with the encrypted A.
+ * Defined in NIST SP 800-38F section 6.1
  */
-void mbedtls_nist_kw_init(mbedtls_nist_kw_context *ctx)
-{
-    memset(ctx, 0, sizeof(mbedtls_nist_kw_context));
-}
-
-int mbedtls_nist_kw_setkey(mbedtls_nist_kw_context *ctx,
-                           mbedtls_cipher_id_t cipher,
-                           const unsigned char *key,
-                           unsigned int keybits,
-                           const int is_wrap)
+static void calc_a_xor_t(unsigned char A[KW_SEMIBLOCK_LENGTH], uint64_t t)
 {
-    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
-    const mbedtls_cipher_info_t *cipher_info;
-
-    cipher_info = mbedtls_cipher_info_from_values(cipher,
-                                                  keybits,
-                                                  MBEDTLS_MODE_ECB);
-    if (cipher_info == NULL) {
-        return MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
-    }
-
-    if (mbedtls_cipher_info_get_block_size(cipher_info) != 16) {
-        return MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
-    }
-
-    /*
-     * SP 800-38F currently defines AES cipher as the only block cipher allowed:
-     * "For KW and KWP, the underlying block cipher shall be approved, and the
-     *  block size shall be 128 bits. Currently, the AES block cipher, with key
-     *  lengths of 128, 192, or 256 bits, is the only block cipher that fits
-     *  this profile."
-     *  Currently we don't support other 128 bit block ciphers for key wrapping,
-     *  such as Camellia and Aria.
-     */
-    if (cipher != MBEDTLS_CIPHER_ID_AES) {
-        return MBEDTLS_ERR_CIPHER_FEATURE_UNAVAILABLE;
+    size_t i = 0;
+    for (i = 0; i < sizeof(t); i++) {
+        A[i] ^= (t >> ((sizeof(t) - 1 - i) * 8)) & 0xff;
     }
+}
 
-    mbedtls_cipher_free(&ctx->cipher_ctx);
+static int verify_input(psa_key_attributes_t *attributes, mbedtls_svc_key_id_t *key){
+    int ret = PSA_SUCCESS;
 
-    if ((ret = mbedtls_cipher_setup(&ctx->cipher_ctx, cipher_info)) != 0) {
-        return ret;
-    }
+    ret = psa_get_key_attributes(*key, attributes);
 
-    if ((ret = mbedtls_cipher_setkey(&ctx->cipher_ctx, key, keybits,
-                                     is_wrap ? MBEDTLS_ENCRYPT :
-                                     MBEDTLS_DECRYPT)
-         ) != 0) {
+    if (ret != PSA_SUCCESS) {
         return ret;
     }
 
-    return 0;
-}
-
-/*
- * Free context
- */
-void mbedtls_nist_kw_free(mbedtls_nist_kw_context *ctx)
-{
-    if (ctx == NULL) {
-        return;
+    /*
+     * Currently NIST KW only supports PSA_KEY_TYPE_AES, so verify this is
+     * set in the key attributes.
+     */
+    if (psa_get_key_type(attributes) != PSA_KEY_TYPE_AES) {
+        return PSA_ERROR_NOT_PERMITTED;
```

Comment:
As [previously discussed](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/180#discussion_r2075846139), this isn't some kind of policy violation, so the error code should not be `NOT_PERMITTED`. Either `INVALID_ARGUMENT` or `NOT_SUPPORTED` would be ok.